benchmark/csharp/Tpch/TpchFunctionalQueries.cs

```diff
@@ -202,6 +226,40 @@ internal void Q8()
               .Show();
         }
 
+        internal void Q8v()
+        {
+            Func<Column, Column> getYear = Udf<string, string>(x => x.Substring(0, 4));
```

Comment:
No, it doesn't.  I implemented both getYear and isBrazil as VectorUdfs, and it seems to slightly degrade the perf:

|                      Run|   With  Udf|   With VectorUdf|
|---------------------------- |----------:|----------:|
| #1 | 28.845 s | 29.246 s |
| #2 | 28.900 s | 29.527 s |
| #3 | 29.143 s | 29.328 s |

Intuitively, I don't think it will help much because these UDFs are not called with a lot of data at once, but instead they are called many times with only a few records at a time. Looking at a log output:

<details>
    <summary>Click to expand</summary>

        [Info] [DaemonWorker] Started .NET DaemonServer with port 39275.
        [Info] [DaemonWorker] New connection accepted for TaskRunner [1]
        [Info] [DaemonWorker] Pool snapshot: [NumThreads:1], [NumTaskRunners:1]
        [Info] [TaskRunner] [1] Starting with ReuseSocket[True].
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:827630
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:822015
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:820344
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:820366
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:820475
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:820506
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:820600
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:249279
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:43
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:31
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:36
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:32
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:54
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:27
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:32
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:33
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:55
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:59
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:31
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:36
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:30
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:56
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:33
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:58
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:36
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:36
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:60
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:28
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:43
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:62
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:43
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:33
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:55
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:52
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:50
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:34
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:54
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:42
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:55
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:39
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:54
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:33
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:43
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:46
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:48
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:44
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:31
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:49
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:45
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:41
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:53
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:62
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:35
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:40
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:30
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:37
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:51
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:38
        [Debug] [TaskRunner] [1] Received END_OF_STREAM signal.
        [Info] [TaskRunner] [1] Processed a task: readComplete:True, entries:47
</details>

You can see the first 8 processed tasks contain hundreds of thousands of entries. But then after that, it suddenly drops to ~40 entries per invocation. The first 8 tasks are processing the OrderLines (which there are millions of records). But the tasks after that are the joined orders.

If we could get Spark to process more entries per invocation, I think we could speed this query up.

Reply:
I think it's a bit non-intuitive to have a vector UDF for one and regular UDFs for others. Since this Q8v is for vector UDFs, can we convert all UDFs to use VectorUDFs? This way, I think it is easier to benchmark VectorUDF vs. UDF.

Reply:
> it's a bit non-intuitive to have a vector UDF for one and regular UDFs for others.

I'm not sure I fully agree. I would think the typical scenario would be to use the function that makes the most sense for the scenario at hand. I wouldn't expect users to use the same one across the board just because it makes sense in one spot.

I can change it to use all VectorUDFs for this function, if the goal is to compare one-at-a-time UDFs vs. the VectorUDFs across the board. But I thought the goal was to show the fastest way possible to execute the query.

Reply:
I had a slightly different idea in mind (which makes this conversation even more interesting :) ). There are a few reasons to get VectorUDFs across the board, in my opinion:

  1. It allows not only for a one-at-a-time comparison but also allows us to do a holistic comparison to understand the overall benefits for a given workload.
  2. We can use the comparison to actually evaluate whether it is worthwhile to profile a specific case and understand if anything else is needed. Optionally, we can get any other curious mind to look and contribute.
  3. Gives us a framework to suggest to customers who have been asking about this on whether VectorUDFs really help them or not for their use case (ideally, I was expecting this would make for an excellent blog post in the future that talks about when to use VectorUDFs and when not to).

I haven't paid too much attention to the effort required to enable this across the board though so I apologize in advance as I don't mean to rat-hole this discussion. Just thought I'll share my two cents :)

Reply:
If the goal is to do a holistic comparison, then I would think you would want every permutation implemented, correct?

For example, if I had 2 UDFs in my query, and one UDF was faster written as a Vector UDF and the other was faster as a normal UDF, I would want the full table of `Normal/Normal`, `Normal/Vector`, `Vector/Normal`, `Vector/Vector`, to compare the differences. If I was only comparing `Normal/Normal` to `Vector/Vector`, the time differences might wash each other out, and it looks like it doesn't make a difference. When really the `Vector/Normal` permutation is the fastest and the `Normal/Vector` combination is the slowest.

Reply:
I was thinking we could do "all normal UDFs" vs "all VectorUdFs", then analyze to come up with hybrid that will perform the best.

But, for the scope of this PR, I think the current approach is fine. We can sync up on the plan for TPCH benchmark in a separate thread or offline.

Reply:
I agree. We can keep it simple and do all normal vs. all vector to begin with. I do not want to delay this PR so I'll leave it to your judgement.